firestore/src/main/java/com/firebase/ui/firestore/paging/FirestoreDataSource.java

```diff
@@ -0,0 +1,208 @@
+        final PageKey key = params.key;
+
+        // Set loading state
+        mLoadingState.postValue(LoadingState.LOADING_MORE);
+
+        key.getPageQuery(mBaseQuery, params.requestedLoadSize)
+                .get()
+                .addOnSuccessListener(new OnLoadSuccessListener() {
+                    @Override
+                    protected void setResult(@NonNull QuerySnapshot snapshot) {
+                        PageKey nextPage = getNextPageKey(snapshot);
+                        callback.onResult(snapshot.getDocuments(), nextPage);
+                    }
+                })
+                .addOnFailureListener(new OnLoadFailureListener() {
+                    @Override
+                    protected Runnable getRetryRunnable() {
+                        return getRetryLoadAfter(params, callback);
+                    }
+                });
+
+    }
+
+    private PageKey getNextPageKey(@NonNull QuerySnapshot snapshot) {
+        List<DocumentSnapshot> data = snapshot.getDocuments();
+        DocumentSnapshot last = getLast(data);
+
+        return new PageKey(last, null);
+    }
+
+    public LiveData<LoadingState> getLoadingState() {
+        return mLoadingState;
+    }
+
+    public void retry() {
+        LoadingState currentState = mLoadingState.getValue();
+        if (currentState != LoadingState.ERROR) {
+            Log.w(TAG, "retry() not valid when in state: " + currentState);
+            return;
+        }
+
+        if (mRetryRunnable == null) {
+            Log.w(TAG, "retry() called with no eligible retry runnable.");
+            return;
+        }
+
+        mRetryRunnable.run();
+    }
+
+    @Nullable
+    private DocumentSnapshot getLast(@NonNull List<DocumentSnapshot> data) {
+        if (data.isEmpty()) {
+            return null;
+        } else {
+            return data.get(data.size() - 1);
+        }
+    }
+
+    @NonNull
+    private Runnable getRetryLoadAfter(@NonNull final LoadParams<PageKey> params,
+                                       @NonNull final LoadCallback<PageKey, DocumentSnapshot> callback) {
+        return new Runnable() {
+            @Override
+            public void run() {
+                loadAfter(params, callback);
+            }
+        };
+    }
+
+    @NonNull
+    private Runnable getRetryLoadInitial(@NonNull final LoadInitialParams<PageKey> params,
+                                         @NonNull final LoadInitialCallback<PageKey, DocumentSnapshot> callback) {
+        return new Runnable() {
+            @Override
+            public void run() {
+                loadInitial(params, callback);
+            }
+        };
+    }
+
+    /**
+     * Success listener that sets success state and nullifies the retry runnable.
+     */
+    private abstract class OnLoadSuccessListener implements OnSuccessListener<QuerySnapshot> {
+
+        @Override
+        public void onSuccess(QuerySnapshot snapshots) {
```

Comment:
Nit: `snapshot`

Reply:
FYI the reason I keep using `snapshots` is because the IDE does that by default, probably because `QuerySnapshot implements Iterable<QueryDocumentSnapshot>`

Reply:
Haha, good ol' IntelliJ! 😉